server/app/assets/javascripts/admin_predicate_configuration.ts

```diff
@@ -0,0 +1,356 @@
+    ) as HTMLSelectElement
+    const operatorValue =
+      operatorDropdown.options[operatorDropdown.options.selectedIndex].value
+
+    const valueInputs = assertNotNull(
+      document
+        ?.querySelector('#predicate-config-value-row-container')
+        ?.querySelectorAll(`[data-question-id="${questionId}"] input`),
+    )
+
+    for (const valueInput of Array.from(valueInputs)) {
+      switch (selectedScalarType.toUpperCase()) {
+        case 'STRING':
+          if (selectedScalarValue.toUpperCase() === 'EMAIL') {
+            // Need to look at the selected scalar *value* for email since the type is just a
+            // string, but emails have a special type in HTML inputs.
+            valueInput.setAttribute('type', 'email')
+            break
+          }
+          valueInput.setAttribute('type', 'text')
+          break
+        case 'CURRENCY_CENTS':
+          if (
+            operatorValue.toUpperCase() === 'IN' ||
+            operatorValue.toUpperCase() === 'NOT_IN'
+          ) {
+            // IN and NOT_IN operate on lists of longs, which must be entered as a comma-separated list
+            valueInput.setAttribute('type', 'text')
+          } else {
+            valueInput.setAttribute('step', '.01')
+            valueInput.setAttribute('placeholder', '$0.00')
+            valueInput.setAttribute('type', 'number')
+          }
+          break
+        case 'LONG':
+          if (
+            operatorValue.toUpperCase() === 'IN' ||
+            operatorValue.toUpperCase() === 'NOT_IN'
+          ) {
+            // IN and NOT_IN operate on lists of longs, which must be entered as a comma-separated list
+            valueInput.setAttribute('type', 'text')
+          } else {
+            valueInput.setAttribute('step', '1')
+            valueInput.setAttribute('placeholder', '0')
+            valueInput.setAttribute('type', 'number')
+          }
+          break
+        case 'DATE':
+          valueInput.setAttribute('type', 'date')
+          break
+        default:
+          valueInput.setAttribute('type', 'text')
+      }
+    }
+  }
+
+  /** Add a value row by duplicating the last row in the list and updating its group IDs. */
+  predicateAddValueRow(event: Event) {
+    event.preventDefault()
+    event.stopPropagation()
+
+    const currentRows = document.querySelectorAll('.predicate-config-value-row')
+    const templateRow = currentRows[currentRows.length - 1]
+    const newRow = templateRow.cloneNode(true) as HTMLElement
+
+    newRow.querySelectorAll('input').forEach((el: HTMLInputElement) => {
+      if (el.type === 'checkbox' || el.type === 'radio') {
+        el.checked = false
+      } else {
+        el.value = ''
+      }
+
+      const groupNumString = assertNotNull(el.name.match(/group-(\d+)/))[1]
+
+      let groupNum = parseInt(groupNumString, 10)
+      el.name = el.name.replace(/group-\d+/, `group-${++groupNum}`)
+      const newId = `${el.id}-${groupNum}`
+      el.id = newId
```

Comment:
> Does this not effectively just add a the groupNum as a suffix each time

Yes. The pattern isn't really important, the goal of setting the ID is to ensure the ID is unique and referenced by the input's label.

Reply:
Please comment that, that you're just trying to make unique ids.  It looks odd/wrong otherwise.

Do these ids end up in the form data such that the server is having to interpret this cascading naming pattern?

Isn't the question ID on the input though? can we not do QID-GroupID just as easy?

Reply:
> Do these ids end up in the form data such that the server is having to interpret this cascading naming pattern?

No. The `name` and `value` attributes of form controls specify the key/value pairs sent to the server. https://developer.mozilla.org/en-US/docs/Learn/Forms/Your_first_form#sending_form_data_to_your_web_server

> Isn't the question ID on the input though? can we not do QID-GroupID just as easy?

The original IDs are generated using UUIDs on the server to ensure uniqueness on the page. Reusing those as substrings in the new IDs helps keep them unique. I'll mention that in the comment here.